containers/src/Data/IntMap/Internal.hs

```diff
@@ -728,6 +729,42 @@ unsafeFindMax Nil = Nothing
 unsafeFindMax (Tip ky y) = Just (ky, y)
 unsafeFindMax (Bin _ _ _ r) = unsafeFindMax r
 
+{--------------------------------------------------------------------
+  Disjoint
+--------------------------------------------------------------------}
+-- | /O(n+m)/. Check whether the key sets of two maps are disjoint
+-- (i.e. their 'intersection' is empty).
+--
+-- > disjoint (fromList [(2,'a')]) (fromList [(1,()), (3,())])   == True
+-- > disjoint (fromList [(2,'a')]) (fromList [(1,'a'), (2,'b')]) == False
+-- > disjoint (fromList [])        (fromList [])                 == True
+--
+-- > disjoint a b == null (intersection a b)
+--
+-- @since UNRELEASED
+disjoint :: IntMap a -> IntMap b -> Bool
+disjoint Nil _ = True
+disjoint _ Nil = True
+disjoint (Tip kx _) (Tip ky _) = kx /= ky
+disjoint (Tip k _) (Bin p m l r) | nomatch k p m = True
+                                 | zero k m      = notMember k l
+                                 | otherwise     = notMember k r
+disjoint (Bin p m l r) (Tip k _) | nomatch k p m = True
+                                 | zero k m      = notMember k l
+                                 | otherwise     = notMember k r
```

Comment:
I think you can collapse all the `Tip` cases into two.

```haskell
disjoint (Tip kx _) ys = notMember kx ys
disjoint xs (Tip ky _) = notMember ky xs
```

Reply:
:+1:

I've also changed the order of the alternatives so the more likely ones come first.